src/ParallelAlgorithms/Amr/Actions/UpdateAmrDecision.hpp

```diff
@@ -0,0 +1,129 @@
+#include "Domain/Structure/ElementId.hpp"
+#include "Domain/Structure/Neighbors.hpp"
+#include "Domain/Tags.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Utilities/ErrorHandling/Error.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/PrettyType.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace amr::Actions {
+/// \brief Given the AMR decision of a neighboring Element, potentially update
+/// the AMR decision of the target Element.
+///
+/// DataBox:
+/// - Uses:
+///   * domain::Tags::Element<volume_dim>
+/// - Modifies:
+///   * amr::domain::Tags::NeighborFlags
+///   * amr::domain::Tags::Flags (if AMR decision is updated)
+///
+/// Invokes:
+/// - amr::Actions::UpdateAmrDecision on all neighboring Element%s (if AMR
+///   decision is updated)
+///
+/// \details This Element calls amr::domain::update_amr_decision to see if its
+/// AMR decision needs to be updated.  If it does, the Element will call
+/// amr::Actions::UpdateAmrDecision on its neighbors.
+///
+struct UpdateAmrDecision {
+  template <typename ParallelComponent, typename DbTagList,
+            typename Metavariables, typename ArrayIndex>
+  static void apply(
+      db::DataBox<DbTagList>& box, Parallel::GlobalCache<Metavariables>& cache,
+      const ArrayIndex& /*array_index*/,
+      const ElementId<Metavariables::volume_dim>& neighbor_id,
+      const std::array<amr::domain::Flag, Metavariables::volume_dim>&
+          neighbor_amr_flags) {
+    constexpr size_t volume_dim = Metavariables::volume_dim;
+
+    // Actions can be executed in any order.  Therefore we need to check:
+    // - If we have evaluated our own AMR decision.  If not, just store
+    //   the neighbor flags and return.
+    std::array<amr::domain::Flag, Metavariables::volume_dim>
+        my_current_amr_flags = get<amr::domain::Tags::Flags<volume_dim>>(box);
+    if (amr::domain::Flag::Undefined == my_current_amr_flags[0]) {
+      db::mutate<amr::domain::Tags::NeighborFlags<volume_dim>>(
+          make_not_null(&box),
+          [&neighbor_id,
+           &neighbor_amr_flags](const gsl::not_null<std::unordered_map<
+                                    ElementId<volume_dim>,
+                                    std::array<amr::domain::Flag, volume_dim>>*>
+                                    amr_flags_of_neighbors) {
+            auto& my_neighbor_amr_flags =
+                (*amr_flags_of_neighbors)[neighbor_id];
+            for (size_t d = 0; d < volume_dim; ++d) {
+              my_neighbor_amr_flags[d] = neighbor_amr_flags[d];
+            }
+          });
+      return;
+    }
+    // - If the neighbor flags we just received have higher priority than
+    //   the flags we already received.  If not, then return.
+    const auto& my_neighbors_amr_flags =
+        get<amr::domain::Tags::NeighborFlags<volume_dim>>(box);
+    if (1 == my_neighbors_amr_flags.count(neighbor_id)) {
+      bool received_old_decision = true;
```

Comment:
I don't understand the name of this bool. What makes the decision "old"? Isn't this checking that if we need to update our decision?

Reply:
bool is gone